### ISSUE TITLE
Change to docker 1.7 events breaks logspout

### ISSUE THREAD

**Issue by phobologic** [gliderlabs/logspout#106]
Due to this bug:

https://github.com/docker/docker/issues/12848#issuecomment-99530374

I have had to use docker 1.7 in my project, or else face running out of memory rapidly.

While implementing logspout (via the docker container, gliderlabs/logspout:master) with docker 1.7, I noticed that a restart of logspout would cause it to crash repeatedly at this line with containers that weren't listed in `docker ps`:

https://github.com/gliderlabs/logspout/blob/master/router/pump.go#L124

It took a bit, but finally I tracked it down to this change to the events code in docker 1.7:

https://github.com/docker/docker/blob/master/daemon/events/events.go#L28

Basically it looks like the events endpoint now returns the last 64 events and then starts tailing the events.  The logspout daemon was crashing because it was working through the past 64 events, and it hit the event that said that the old logspout daemon had just come up.

I want to go ahead and make an attempt at fixing this issue, but first I figured I'd get some people's take on what the behavior should be.

**Comment by dreamcat4** [comment]
Maybe try something like this?

```sh
_start_time="$(date +%s)"
docker events --since=$_start_time
```

**Comment by progrium** [comment]
Docker events has always done this so I'm not sure if it's a default that
changed or what, but I suspect this will need more investigation. Let me
know if you find something that fixes it. In terms of what the behavior
should be, we only use events for new events since we iterate over existing
containers separately. So we really don't care about old events at all.

On Fri, Jun 5, 2015 at 12:27 PM, Dreamcat4 <notifications@github.com> wrote:

> Maybe try something like this?
>
> _start_time="$(date +%s)"
> docker events --since=$_start_time
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/gliderlabs/logspout/issues/106#issuecomment-109370114>
> .
>



-- 
Jeff Lindsay
http://progrium.com

**Comment by dreamcat4** [comment]
My only point was that by using `--since=$now` flag it should not be printing the last 64 event messages before it is starting to `tail`. Which is what you need to avoid the new behavior isn't it?

I also will be affected by this in the same way. Yet have not upgraded to 1.7. Therefore cannot check myself for sure if that will have the desired effect (but it aught to!)

:)

**Comment by phobologic** [comment]
It doesn't look like it has - 1.6 used to just return events from the moment you made the request, now it returns the last 64 then starts tailing.

I opened a bug w/ Docker, since the behavior is kind of tricky to work around:

https://github.com/docker/docker/issues/13753

They've verified that it's changed.  It'd be great if they could provide a way to go back to the old functionality at least.

**Comment by phobologic** [comment]
Oh, sorry, I think I'm getting the comment stream mixed up :)

**Comment by progrium** [comment]
If you can't go back to the old behavior that is definitely a bug.

**Comment by phobologic** [comment]
Just thought I'd share - it looks like a fix is incoming in 1.7.  I'll test it as soon as it's out:

https://github.com/docker/docker/pull/13754

**Comment by phobologic** [comment]
https://github.com/docker/docker/issues/13753 has been closed and https://github.com/docker/docker/issues/13754 has been merged.  I'm going to test it as soon as there is a build (not sure how long that takes, may already be there).

**Comment by phobologic** [comment]
I've tested the merge and everything is working great.  Closing this.  Thanks!
